src/lib.rs

```diff
@@ -19,33 +19,88 @@
 //! ## Usage
 //! ```rust
 //! use aes_ccm::CcmMode;
+//! use aes_ccm::aead::{NewAead, Aead, Payload};
+//! use aes_ccm::aead::generic_array::{GenericArray, typenum::U8};
 //!
 //! let key = [
 //!     0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA,
 //!     0xCB, 0xCC, 0xCD, 0xCE, 0xCF,
 //! ];
+//!
+//! // `U8` represents the tag size as a `typenum` unsigned (8-bytes here)
+//! let ccm = CcmMode::<U8>::new(key.into());
+//!
 //! let nonce = [
 //!     0x00, 0x00, 0x00, 0x03, 0x02, 0x01, 0x00, 0xA0, 0xA1, 0xA2, 0xA3,
 //!     0xA4, 0xA5,
 //! ];
-//! let ccm = CcmMode::new(&key, nonce, 8).unwrap();
-//!
-//! let mut ciphertext_buf = [0u8; 50];
-//! let payload = [
+//! let msg = [
 //!     0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10, 0x11, 0x12,
 //!     0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1D,
 //!     0x1E,
 //! ];
 //! let associated_data = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07];
 //! let ciphertext = ccm
-//!     .generate_encrypt(&mut ciphertext_buf, &associated_data, &payload)
+//!     .encrypt(&nonce.into(), Payload { aad: &associated_data, msg: &msg })
 //!     .unwrap();
 //!
-//! let mut plaintext_buf = [0u8; 50];
 //! let plaintext = ccm
-//!     .decrypt_verify(&mut plaintext_buf, &associated_data, &ciphertext)
+//!     .decrypt(&nonce.into(), Payload { aad: &associated_data, msg: &ciphertext })
 //!     .unwrap();
-//! assert_eq!(&payload, plaintext);
+//! assert_eq!(&msg[..], plaintext.as_slice());
+//! ```
+//!
+//! ## In-place Usage (eliminates `alloc` requirement)
+//! This crate has an optional `alloc` feature which can be disabled in e.g.
+//! microcontroller environments that don't have a heap.
+//!
+//! The [`Aead::encrypt_in_place`] and [`Aead::decrypt_in_place`]
+//! methods accept any type that impls the [`aead::Buffer`] trait which
+//! contains the plaintext for encryption or ciphertext for decryption.
+//!
+//! Note that if you enable the `heapless` feature of this crate,
+//! you will receive an impl of `aead::Buffer` for [`heapless::Vec`]
+//! (re-exported from the `aead` crate as `aead::heapless::Vec`),
+//! which can then be passed as the `buffer` parameter to the in-place encrypt
+//! and decrypt methods:
+//!
+//! ```
+//! use aes_ccm::CcmMode;
+//! use aes_ccm::aead::{NewAead, Aead, Payload};
+//! use aes_ccm::aead::generic_array::{GenericArray, typenum::{U8, U128}};
+//! use aead::heapless::Vec;
+//!
+//! let key = [
+//!     0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA,
+//!     0xCB, 0xCC, 0xCD, 0xCE, 0xCF,
+//! ];
+//!
+//! // `U8` represents the tag size as a `typenum` unsigned (8-bytes here)
+//! let ccm = CcmMode::<U8>::new(key.into());
```

Comment:
We could define some type aliases to make these not require the turbofish, e.g.

```rust
pub type CcmMode8 = CcmMode<U8>;
```

Reply:
I don't think that's necessary, it's not that hard to use with the turbofish. And for people who dislike this syntax, there's still the option of doing it with
```rust
let ccm: AesCcm<U8> = CcmMode::new(key.into());
```
right?

Reply:
Yep (or defining their own type aliases)